File: app/src/main/java/dev/ragnarok/fenrir/api/interfaces/IAudioApi.java
package dev.ragnarok.fenrir.api.interfaces;

import androidx.annotation.CheckResult;
import androidx.annotation.NonNull;

import java.util.Collection;
import java.util.List;

import dev.ragnarok.fenrir.api.model.AccessIdPair;
import dev.ragnarok.fenrir.api.model.IdPair;
import dev.ragnarok.fenrir.api.model.Items;
import dev.ragnarok.fenrir.api.model.VKApiAudio;
import dev.ragnarok.fenrir.api.model.VKApiAudioCatalog;
import dev.ragnarok.fenrir.api.model.VKApiAudioPlaylist;
import dev.ragnarok.fenrir.api.model.VkApiArtist;
import dev.ragnarok.fenrir.api.model.VkApiLyrics;
import dev.ragnarok.fenrir.api.model.response.AddToPlaylistResponse;
import dev.ragnarok.fenrir.api.model.response.CatalogResponse;
import dev.ragnarok.fenrir.api.model.response.ServicePlaylistResponse;
import dev.ragnarok.fenrir.api.model.server.VkApiAudioUploadServer;
import dev.ragnarok.fenrir.model.Audio;
import io.reactivex.rxjava3.core.Single;


public interface IAudioApi {

    @CheckResult
    Single<int[]> setBroadcast(IdPair audio, Collection<Integer> targetIds);

    @CheckResult
    Single<Items<VKApiAudio>> search(String query, Boolean autoComplete, Boolean lyrics,
                                     Boolean performerOnly, Integer sort, Boolean searchOwn,
                                     Integer offset, Integer count);

    @CheckResult
    Single<Items<VkApiArtist>> searchArtists(String query, Integer offset, Integer count);

    @CheckResult
    Single<Items<VKApiAudioPlaylist>> searchPlaylists(String query, Integer offset, Integer count);

    @CheckResult
    Single<VKApiAudio> restore(int audioId, Integer ownerId);

    @CheckResult
    Single<Boolean> delete(int audioId, int ownerId);

    @CheckResult
    Single<Integer> edit(int ownerId, int audioId, String artist, String title, String text);

    @CheckResult
    Single<Integer> add(int audioId, int ownerId, Integer groupId, String accessKey);

    @CheckResult
    Single<VKApiAudioPlaylist> createPlaylist(int ownerId, String title, String description);

    @CheckResult
    Single<Integer> editPlaylist(int ownerId, int playlist_id, String title, String description);

    @CheckResult
    Single<Integer> removeFromPlaylist(int ownerId, int playlist_id, Collection<AccessIdPair> audio_ids);

    @CheckResult
    Single<List<AddToPlaylistResponse>> addToPlaylist(int ownerId, int playlist_id, Collection<AccessIdPair> audio_ids);

    @CheckResult
    Single<Integer> reorder(int ownerId, int audio_id, Integer before, Integer after);

    @CheckResult
    Single<Integer> trackEvents(String events);

    @CheckResult
    Single<Items<VKApiAudio>> get(Integer playlist_id, Integer ownerId,
                                  Integer offset, Integer count, String accessKey);

    @CheckResult
    Single<Items<VKApiAudio>> getAudiosByArtist(String artist_id, Integer offset, Integer count);

    @CheckResult
    Single<List<VKApiAudio>> getPopular(Integer foreign,
                                        Integer genre, Integer count);

    @CheckResult
    Single<Integer> deletePlaylist(int playlist_id, int ownerId);

    @CheckResult
    Single<VKApiAudioPlaylist> followPlaylist(int playlist_id, int ownerId, String accessKey);

    @CheckResult
    Single<VKApiAudioPlaylist> clonePlaylist(int playlist_id, int ownerId);

    @CheckResult
    Single<VKApiAudioPlaylist> getPlaylistById(int playlist_id, int ownerId, String accessKey);

    @CheckResult
    Single<Items<VKApiAudio>> getRecommendations(Integer audioOwnerId, Integer count);

    @CheckResult
    Single<Items<VKApiAudio>> getRecommendationsByAudio(String audio, Integer count);

    @CheckResult
    Single<List<VKApiAudio>> getById(@NonNull List<Audio> audios);

    @CheckResult
    Single<List<VKApiAudio>> getByIdOld(@NonNull List<Audio> audios);

    @CheckResult
    Single<VkApiLyrics> getLyrics(int lyrics_id);

    @CheckResult
    Single<Items<VKApiAudioPlaylist>> getPlaylists(int owner_id, int offset, int count);

    @CheckResult
    Single<ServicePlaylistResponse> getPlaylistsCustom(String code);

    @CheckResult
    Single<Items<VKApiAudioCatalog>> getCatalog(String artist_id, String query);

    @CheckResult
    Single<CatalogResponse> getCatalogBlockById(String block_id, String start_from);

    @CheckResult
    Single<VkApiAudioUploadServer> getUploadServer();

    @CheckResult
    Single<VKApiAudio> save(String server, String audio, String hash, String artist, String title);
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/api/interfaces/IAudioApi.kt
package dev.ragnarok.fenrir.api.interfaces

import androidx.annotation.CheckResult
import dev.ragnarok.fenrir.api.model.*
import dev.ragnarok.fenrir.api.model.catalog_v2_audio.VKApiCatalogV2BlockResponse
import dev.ragnarok.fenrir.api.model.catalog_v2_audio.VKApiCatalogV2ListResponse
import dev.ragnarok.fenrir.api.model.catalog_v2_audio.VKApiCatalogV2SectionResponse
import dev.ragnarok.fenrir.api.model.response.AddToPlaylistResponse
import dev.ragnarok.fenrir.api.model.response.ServicePlaylistResponse
import dev.ragnarok.fenrir.api.model.server.VKApiAudioUploadServer
import dev.ragnarok.fenrir.model.Audio
import io.reactivex.rxjava3.core.Single

interface IAudioApi {
    @CheckResult
    fun setBroadcast(audio: AccessIdPair, targetIds: Collection<Long>): Single<List<Int>>

    @CheckResult
    fun search(
        query: String?, autoComplete: Boolean?, lyrics: Boolean?,
        performerOnly: Boolean?, sort: Int?, searchOwn: Boolean?,
        offset: Int?, count: Int?
    ): Single<Items<VKApiAudio>>

    @CheckResult
    fun searchArtists(query: String?, offset: Int?, count: Int?): Single<Items<VKApiArtist>>

    @CheckResult
    fun searchPlaylists(
        query: String?,
        offset: Int?,
        count: Int?
    ): Single<Items<VKApiAudioPlaylist>>

    @CheckResult
    fun restore(audioId: Int, ownerId: Long?): Single<VKApiAudio>

    @CheckResult
    fun delete(audioId: Int, ownerId: Long): Single<Boolean>

    @CheckResult
    fun edit(
        ownerId: Long,
        audioId: Int,
        artist: String?,
        title: String?,
        text: String?
    ): Single<Int>

    @CheckResult
    fun add(audioId: Int, ownerId: Long, groupId: Long?, accessKey: String?): Single<Int>

    @CheckResult
    fun createPlaylist(
        ownerId: Long,
        title: String?,
        description: String?
    ): Single<VKApiAudioPlaylist>

    @CheckResult
    fun editPlaylist(
        ownerId: Long,
        playlist_id: Int,
        title: String?,
        description: String?
    ): Single<Int>

    @CheckResult
    fun removeFromPlaylist(
        ownerId: Long,
        playlist_id: Int,
        audio_ids: Collection<AccessIdPair>
    ): Single<Int>

    @CheckResult
    fun addToPlaylist(
        ownerId: Long,
        playlist_id: Int,
        audio_ids: Collection<AccessIdPair>
    ): Single<List<AddToPlaylistResponse>>

    @CheckResult
    fun reorder(ownerId: Long, audio_id: Int, before: Int?, after: Int?): Single<Int>

    @CheckResult
    fun trackEvents(events: String?): Single<Int>

    @CheckResult
    operator fun get(
        playlist_id: Int?, ownerId: Long?,
        offset: Int?, count: Int?, accessKey: String?
    ): Single<Items<VKApiAudio>>

    @CheckResult
    fun getAudiosByArtist(
        artist_id: String?,
        offset: Int?,
        count: Int?
    ): Single<Items<VKApiAudio>>

    @CheckResult
    fun getPopular(
        foreign: Int?,
        genre: Int?, count: Int?
    ): Single<List<VKApiAudio>>

    @CheckResult
    fun deletePlaylist(playlist_id: Int, ownerId: Long): Single<Int>

    @CheckResult
    fun followPlaylist(
        playlist_id: Int,
        ownerId: Long,
        accessKey: String?
    ): Single<VKApiAudioPlaylist>

    @CheckResult
    fun clonePlaylist(playlist_id: Int, ownerId: Long): Single<VKApiAudioPlaylist>

    @CheckResult
    fun getPlaylistById(
        playlist_id: Int,
        ownerId: Long,
        accessKey: String?
    ): Single<VKApiAudioPlaylist>

    @CheckResult
    fun getRecommendations(audioOwnerId: Long?, count: Int?): Single<Items<VKApiAudio>>

    @CheckResult
    fun getRecommendationsByAudio(audio: String?, count: Int?): Single<Items<VKApiAudio>>

    @CheckResult
    fun getById(audios: List<Audio>): Single<List<VKApiAudio>>

    @CheckResult
    fun getByIdOld(audios: List<Audio>): Single<List<VKApiAudio>>

    @CheckResult
    fun getLyrics(lyrics_id: Int): Single<VKApiLyrics>

    @CheckResult
    fun getPlaylists(owner_id: Long, offset: Int, count: Int): Single<Items<VKApiAudioPlaylist>>

    @CheckResult
    fun getPlaylistsCustom(code: String?): Single<ServicePlaylistResponse>

    @get:CheckResult
    val uploadServer: Single<VKApiAudioUploadServer>

    @CheckResult
    fun save(
        server: String?,
        audio: String?,
        hash: String?,
        artist: String?,
        title: String?
    ): Single<VKApiAudio>

    @CheckResult
    fun getCatalogV2Sections(
        owner_id: Long, artist_id: String?, url: String?, query: String?, context: String?
    ): Single<VKApiCatalogV2ListResponse>

    @CheckResult
    fun getCatalogV2Section(
        section_id: String,
        start_from: String?
    ): Single<VKApiCatalogV2SectionResponse>

    @CheckResult
    fun getCatalogV2BlockItems(
        block_id: String, start_from: String?
    ): Single<VKApiCatalogV2BlockResponse>

    @CheckResult
    fun getArtistById(
        artist_id: String
    ): Single<ArtistInfo>
}